File: Lab9/spacecraft.py
import os

class Spacecraft:
    def __init__(self, name, engine_type, engine_power, nav_system_type, ammo_left, ammo_right):
        self.name = name
        self.engine = self.Engine(engine_type, engine_power)
        self.navigation_system = self.NavigationSystem(nav_system_type)
        self.left_turret = self.Turret(ammo_left)
        self.right_turret = self.Turret(ammo_right)

    def launch(self):
        pass

    def navigate_to(self, destination):
        pass

    def shoot(self, is_left_shoot, is_right_shoot):
        pass

    def stop(self):
        pass

    def print_status(self):
        pass

    def log(self, message):
        log_file_path = "spacecraft_log.txt"
        with open(log_file_path, "a") as file:
            file.write(message + "\n")

    class Engine:
        def __init__(self, engine_type, power):
            self.type = engine_type
            self.power = power

        def get_type(self):
            return self.type

        def get_power(self):
            return self.power

    class NavigationSystem:
        def __init__(self, nav_type):
            self.type = nav_type

        def get_type(self):
            return self.type

    class Turret:
        def __init__(self, ammo):
            self.ammo = ammo

        def get_ammo(self):
            return self.ammo

        def set_ammo(self, ammo):
            self.ammo = ammo

File: Lab3/src/Spacecraft.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

public abstract class Spacecraft {
    protected String name;
    protected Engine engine;
    protected NavigationSystem navigationSystem;
    protected Turrel LeftTurrel;
    protected Turrel RightTurrel;


    // Конструктори
    public Spacecraft(String name, Engine engine, NavigationSystem navigationSystem, Turrel LeftTurrel, Turrel RightTurrel) {
        this.name = name;
        this.engine = engine;
        this.navigationSystem = navigationSystem;
        this.LeftTurrel = LeftTurrel;
        this.RightTurrel = RightTurrel;
    }

    public Spacecraft(String name, String engineType, int enginePower, String navSystemType, int ammoLeft, int ammoRight) {
        this.name = name;
        this.engine = new Engine(engineType, enginePower);
        this.navigationSystem = new NavigationSystem(navSystemType);
        this.LeftTurrel = new Turrel(ammoLeft);
        this.RightTurrel = new Turrel(ammoRight);
    }

    // Abstract methods
    public abstract void launch();

    public abstract void navigateTo(String destination);

    public abstract void shoot(boolean IsLeftShoot, boolean IsRightShoot);

    public abstract void stop();

    public abstract void printStatus();

    private void log(String message) {
        try (FileWriter fileWriter = new FileWriter("spacecraft_log.txt", true);
             PrintWriter printWriter = new PrintWriter(fileWriter)) {
            printWriter.println(message);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    protected class Engine {
        private String type;
        private int power;

        public Engine(String type, int power) {
            this.type = type;
            this.power = power;
        }

        public String getType() {
            return type;
        }

        public int getPower() {
            return power;
        }
    }

    protected class NavigationSystem {
        private String type;

        public NavigationSystem(String type) {
            this.type = type;
        }
        public String getType() {
            return type;
        }
    }

    protected class Turrel {
        private int ammo;

        public Turrel(int ammo) {
            this.ammo = ammo;
        }

        public int getAmmo() {
            return ammo;
        }

        public void setAmmo(int Ammo) {
            this.ammo = Ammo;

        }

    }
}